src/config.rs

```diff
@@ -81,27 +120,168 @@ impl<'a, Broker: crate::Broker> Config<'a, Broker> {
         self
     }
 
-    /// Construct a will in the provided buffer.
-    pub fn will_with_buffer(
-        mut self,
-        buf: &'a mut [u8],
-        will: Will<'_>,
-    ) -> Result<Self, crate::ser::Error> {
-        self = self.will_buffer(buf);
-        self.will(will)
-    }
-
     /// Specify the Will message to be sent if the client disconnects.
     ///
     /// # Args
     /// * `will` - The will to use.
     pub fn will(mut self, will: Will<'_>) -> Result<Self, crate::ser::Error> {
-        let Some(WillState::BufferAvailable(buf)) = self.will else {
-            return Err(crate::ser::Error::InsufficientMemory);
+        let will_len = will.serialized_len();
+        let (head, tail) = self.buffer.split_at_mut(will_len);
+        self.buffer = tail;
+        self.will = Some(will.serialize(head)?);
+
+        Ok(self)
+    }
+
+    /// Consume the configuration and split the user-provided memory into the necessary buffers.
+    pub(crate) fn build(self) -> Config<'a, Broker> {
+        let configs = [self.rx_config, self.tx_config, self.session_state_config];
+        let mut buffers = [None, None, None];
+
+        let mut data = self.buffer;
+
+        // First, check for any exact configurations.
+        for (size, buffer) in configs
+            .iter()
+            .zip(buffers.iter_mut())
+            .filter_map(|(config, buf)| {
+                if let Some(BufferConfig::Exactly(size)) = config {
+                    Some((size, buf))
+                } else {
+                    None
+                }
+            })
+        {
+            let (head, tail) = data.split_at_mut(*size);
+            data = tail;
+            buffer.replace(head);
+        }
```

Comment:
This is becoming more like an allocator.

Reply:
It felt 100% like an allocator as I was writing it. I don't know if this is a good or bad thing.

In our case, we have an upper limit on the number of buffers we're splitting, and they never deallocate, so it's more like half an allocator.

Reply:
Perhaps we could steal memory from the `session_state` buffer for the will message instead of this heavy-handed approach of implementing a full allocator?

Reply:
Let's keep it like this and see how it goes.